File: domains/migration/lob_chunk_reader.py
"""
Module: domains/migration/lob_chunk_reader.py
Purpose: Chunked LOB reading via SQL Server SUBSTRING() to avoid OOM on large blobs.
         Reads a single LOB value for a given PK in configurable byte-sized pieces,
         eliminating the need to materialise multi-GB values in Python memory.
Author: Ravi Sharma
Copyright (c) 2026 Ravi Sharma
SPDX-License-Identifier: MIT
"""
from __future__ import annotations

from typing import Any

from shared.kernel.sql_identifier import validate_sql_identifier
from shared.logging.structured_logging import get_logger

logger = get_logger(__name__)

_DEFAULT_CHUNK_SIZE_BYTES = 1024 * 1024  # 1 MB


class LobChunkReader:
    """Reads a single LOB/CLOB column value in bounded-memory chunks.

    Uses SQL Server's ``SUBSTRING(col, offset, chunk_size)`` to stream
    one LOB at a time, avoiding OOM for columns exceeding available RAM.

    Example::

        reader = LobChunkReader(chunk_size_bytes=1024 * 1024)
        data = await reader.read_lob(
            connector, "dbo", "files", "file_id", pk_value=42, col="content"
        )

    The *data* result is a ``bytes`` object assembled from all chunks, or
    a ``str`` if the database returned text chunks, or ``None`` if the column
    value is SQL NULL.
    """

    def __init__(self, chunk_size_bytes: int = _DEFAULT_CHUNK_SIZE_BYTES) -> None:
        self._chunk_size = chunk_size_bytes

    @property
    def chunk_size_bytes(self) -> int:
        return self._chunk_size

    async def read_lob(
        self,
        connector: Any,
        schema: str,
        table: str,
        pk_col: str,
        pk_value: Any,
        col: str,
    ) -> bytes | str | None:
        """Read *col* for the row identified by *pk_value* in chunks.

        Returns the fully assembled LOB value, or ``None`` if the column
        contains SQL NULL.

        Args:
            connector:  SQL Server connector with an ``execute(sql, params)``
                        method (parameterised; params as a dict or positional).
            schema:     SQL Server schema name (e.g. ``"dbo"``).
            table:      Table name.
            pk_col:     Primary-key column that uniquely identifies the row.
            pk_value:   The PK value used to filter the row (passed as a
                        query parameter — never string-interpolated).
            col:        LOB column name to read.

        Returns:
            Assembled ``bytes`` or ``str``, or ``None`` for a NULL column.
        """
        validate_sql_identifier(schema, "schema")
        validate_sql_identifier(table, "table")
        validate_sql_identifier(pk_col, "pk_col")
        validate_sql_identifier(col, "col")

        # SUBSTRING uses 1-based offsets in SQL Server / T-SQL.
        sql_offset = 1
        chunks_binary: list[bytes] = []
        chunks_text: list[str] = []
        is_binary: bool | None = None

        while True:
            query = (
                f"SELECT SUBSTRING([{col}], ?, ?) AS chunk "
                f"FROM [{schema}].[{table}] "
                f"WHERE [{pk_col}] = ?"
            )
            rows = await connector.execute(query, (sql_offset, self._chunk_size, pk_value))

            if not rows:
                break

            chunk = rows[0].get("chunk")

            # NULL column — signal that the entire LOB is NULL
            if chunk is None:
                if sql_offset == 1:
                    return None
                break

            # Empty bytes / empty string — SUBSTRING past end of value
            if isinstance(chunk, (bytes, bytearray)) and len(chunk) == 0:
                break
            if isinstance(chunk, str) and len(chunk) == 0:
                break

            if is_binary is None:
                is_binary = isinstance(chunk, (bytes, bytearray))

            if is_binary:
                chunks_binary.append(bytes(chunk))
            else:
                chunks_text.append(str(chunk))

            sql_offset += self._chunk_size

        if is_binary is None:
            # No data was ever returned (all None/empty at offset 1)
            return None

        if is_binary:
            return b"".join(chunks_binary)
        return "".join(chunks_text)

File: engine-go/internal/extractor/lob_chunk_reader.go
// Copyright (c) 2026 Ravi Sharma. MIT License.
// SPDX-License-Identifier: MIT

package extractor

import (
	"context"
	"database/sql"
	"fmt"
)

const defaultLOBChunkSizeBytes = 1024 * 1024 // 1 MB — mirrors Python LobChunkReader default

// LobChunkReader reads a single LOB column via SQL Server SUBSTRING() in bounded chunks.
type LobChunkReader struct {
	ChunkSizeBytes int
}

// DefaultLobChunkReader returns a reader with the Python-default 1 MB chunk size.
func DefaultLobChunkReader() *LobChunkReader {
	return &LobChunkReader{ChunkSizeBytes: defaultLOBChunkSizeBytes}
}

// BuildLOBSubstringQuery returns parameterised SUBSTRING SQL for one LOB chunk.
func BuildLOBSubstringQuery(schema, table, col, pkCol string) string {
	return fmt.Sprintf(
		"SELECT SUBSTRING(%s, @P1, @P2) AS chunk FROM %s WHERE %s = @P3",
		quoteIdentMSSQL(col),
		quotedTableMSSQL(schema, table),
		quoteIdentMSSQL(pkCol),
	)
}

// ReadLOB assembles a full LOB value for one row using 1-based SUBSTRING offsets.
func (r *LobChunkReader) ReadLOB(
	ctx context.Context,
	db *sql.DB,
	schema, table, pkCol, col string,
	pkValue any,
) (any, error) {
	if r.ChunkSizeBytes < 1 {
		r.ChunkSizeBytes = defaultLOBChunkSizeBytes
	}
	query := BuildLOBSubstringQuery(schema, table, col, pkCol)

	sqlOffset := 1
	var chunksBin [][]byte
	var chunksText []string
	var isBinary *bool

	for {
		var chunk any
		err := db.QueryRowContext(ctx, query,
			sql.Named("P1", sqlOffset),
			sql.Named("P2", r.ChunkSizeBytes),
			sql.Named("P3", pkValue),
		).Scan(&chunk)
		if err == sql.ErrNoRows {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("read LOB chunk for %s.%s: %w", schema, table, err)
		}
		if chunk == nil {
			if sqlOffset == 1 {
				return nil, nil
			}
			break
		}

		switch v := chunk.(type) {
		case []byte:
			if len(v) == 0 {
				if isBinary == nil {
					return nil, nil
				}
				goto done
			}
			if isBinary == nil {
				b := true
				isBinary = &b
			}
			chunksBin = append(chunksBin, v)
		case string:
			if len(v) == 0 {
				if isBinary == nil {
					return nil, nil
				}
				goto done
			}
			if isBinary == nil {
				b := false
				isBinary = &b
			}
			chunksText = append(chunksText, v)
		default:
			return nil, fmt.Errorf("unexpected LOB chunk type %T", chunk)
		}
		sqlOffset += r.ChunkSizeBytes
	}
done:
	if isBinary == nil {
		return nil, nil
	}
	if *isBinary {
		total := 0
		for _, c := range chunksBin {
			total += len(c)
		}
		out := make([]byte, 0, total)
		for _, c := range chunksBin {
			out = append(out, c...)
		}
		return out, nil
	}
	return joinStrings(chunksText), nil
}

func joinStrings(parts []string) string {
	if len(parts) == 0 {
		return ""
	}
	if len(parts) == 1 {
		return parts[0]
	}
	n := 0
	for _, p := range parts {
		n += len(p)
	}
	b := make([]byte, 0, n)
	for _, p := range parts {
		b = append(b, p...)
	}
	return string(b)
}
